afkak/kafkacodec.py

```diff
@@ -322,7 +406,46 @@ def _decode_message(cls, data, offset):
     ##################
 
     @classmethod
-    def get_response_correlation_id(cls, data):
+    def encode_api_versions_request(
+        cls, client_id: bytes, correlation_id: int, api_version_request: ApiVersionRequest
+    ) -> bytes:
+        """
+        Encode an ApiVersionsRequest. Format::
+
+            ApiVersionsRequest => [ApiVersionRequest]
+                ApiVersionRequest => ApiKey
+        """
+        return cls._encode_message_header(client_id, correlation_id, api_version_request.api_key) + struct.pack(
+            ">i", api_version_request.api_version
+        )
+
+    @classmethod
+    def decode_api_versions_response(cls, data: bytes) -> ApiVersionResponse:
+        """
+        Decode an ApiVersionsResponse. Format::
+
+            ResponseHeader => CorrelationId
+                CorrelationId => int32
+
+            ApiVersionsResponse => ErrorCode [ApiVersions]
+                ErrorCode => int16
+                ApiVersions => ApiKey MinVersion MaxVersion
+                    ApiKey => int16
+                    MinVersion => int16
+                    MaxVersion => int16
+        """
+        ((correlation_id, error_code), cur) = relative_unpack(">ii", data, 0)
+        data = data[2:]  # move past correlation_id and error_code
+
+        api_versions = []
+        while cur < len(data):
```

Comment:
Does this while loop do anything? Shouldn't `iter_unpack` unpack the entire response? If the data is not a multiple of 6 it will just fail so it must consume everything (and if it didn't fail we would just be stuck in this while loop forever because the cursor will never advance).